sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -553,11 +580,36 @@ class CodegenContext {
           }
         """
       addNewFunction(compareFunc, funcCode)
-      s"this.$compareFunc($c1, $c2)"
+      s"$compareFunc($c1, $c2)"
     case schema: StructType =>
-      INPUT_ROW = "i"
-      val comparisons = GenerateOrdering.genComparisons(this, schema)
       val compareFunc = freshName("compareStruct")
+      val compareFields = schema.zipWithIndex.map {
```

Comment:
the previous comparison code does not work in whole-stage codegen, as it mutate the `INPUT_ROW`.